mate3/api.py

```diff
@@ -164,54 +303,56 @@ def set_value(self, field: Field, value: int, port: int = None):
-        )
 
-        # Do the write
-        self.client.write_registers(field_register, prepared_values)
+        # TODO: write in ranges?
+        for device in self._devices.connected_devices:
+            for field_value in device.fields():
+                if field_value.dirty:
+                    if field_value.field.mode not in (Mode.RW, Mode.W):
+                        raise RuntimeError(f"Can't write to field {field_value.name}")
+                    logger.info(f"writing {field_value._value_to_write} to {field_value.name}")
+                    address = device.address + field_value.field.start - 1  # -1 since start is 1-indexed
+                    # TODO: should we check that the device at the given address still has the right port?
+                    registers = field_value.field.to_registers(field_value._value_to_write)
+                    logger.debug(f"Setting register {address} to value {registers}")
+
+                    # # Do the write
+                    self.client.write_registers(address, registers)
+
+
+if __name__ == "__main__":
+
+    import sys
+    import time
+
+    logger.remove()
+    logger.add(sys.stderr, level="INFO")
+
+    cache_path = Path(__file__).parent / ".modbus_client_cache.json"
+    with Mate3Client("192.168.1.12", cache_path=cache_path, cache_only=True) as client:
+        client.read()
+        print("# What's the system name?")
+        print(client.devices.mate3.system_name)
+        print("# Get the battery voltage. Note that it's auto-scaled appropriately.")
+        print(client.devices.fndc.battery_voltage)
+        print("# Get the (raw) values for the same device type on different ports")
+        for port in client.devices.fx_inverters:
+            print(f"FET temp on port {port} = {client.devices.fx_inverters[port].fet_temperature.value}")
+        print("# Read only battery voltage again and check only it's read time was updated but not system name")
+        time.sleep(1)
+        client.read(only=[client.devices.fndc.battery_voltage])
+        print(client.devices.mate3.system_name)
+        print(client.devices.fndc.battery_voltage)
+        print("# Nice. What about modbus fields that aren't implemented?")
+        print(client.devices.mate3.sched_1_ac_mode)
+        print("# Cool. Can we set a new value? Note that we don't need to worry about scaling etc.")
+        volts = client.devices.charge_controller.config.absorb_volts
+        print("Before:", volts)
+        client.devices.charge_controller.config.absorb_volts.value = volts.value + 0.1
+        print("After: ", volts)
+        print("# OK, but what about fun fields like Enums?")
+        new_value = client.devices.charge_controller.config.grid_tie_mode.field.options["Grid Tie Mode disabled"]
+        client.devices.charge_controller.config.grid_tie_mode.value = new_value
+        print("# Finally, write any values that have changed to the device itself - BE CAREFUL!")
+        client.write()
```

Comment:
There is a lot going on here, so I haven't read through it in detail. Thanks for the comments and docstrings through

Reply:
No problems. I tried to convey the various useful things, but it's really wanting a nice bit of documentation / FAQ.